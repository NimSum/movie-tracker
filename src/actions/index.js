export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  movies
})

export const formType = (type) => ({
  type: 'FORM_TYPE',
  form: type
})

export const currentUser = (id, name, email, favorites) => ({
  type: 'CURRENT_USER',
  userInfo: { id, name, email, favorites }
})

export const selectedMovie = (movie) => ({
  type: 'SELECTED_MOVIE',
  movie
})

export const updateFavorites = (favorite) => ({
  type: 'UPDATE_FAVORITES',
  favorite
})