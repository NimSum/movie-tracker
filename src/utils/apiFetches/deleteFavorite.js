export const deleteFavorite = async (userId, movieId) => {
  const url = `http://localhost:3000/api/users/${userId}/favorites/${movieId}`
  const response = await fetch(url, {
    method: 'DELETE',
    headers: {'Content-Type': 'application/json'}
  })
    if (!response.ok) {
      throw new Error('Something went wrong')
    }
  return response.json()
}