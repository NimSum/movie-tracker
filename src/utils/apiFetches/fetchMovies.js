export const fetchMovies = async (url) => {
  const response = await fetch(url);
  if(!response.ok) {
    throw Error("Failed to fetch movies")
  }
  return await response.json();
}