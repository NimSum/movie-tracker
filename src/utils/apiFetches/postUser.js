export const postUser = async (url, body) => {
  const response = await fetch( url, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type' : 'application/json'
    }
  })
  if (!response.ok) {
    throw new Error('Failed to post user data')
  }
  return response.json();
}