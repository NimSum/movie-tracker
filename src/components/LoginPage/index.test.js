import React from 'react';
import { LoginPage, mapStateToProps, mapDispatchToProps } from './index';
import { shallow } from 'enzyme';

import { postUser } from '../../utils/apiFetches/postUser';
import { fetchAnything } from '../../utils/apiFetches/fetchAnything';
import { currentUser, formType } from '../../actions';
import { Redirect } from 'react-router-dom';

jest.mock('../../utils/apiFetches/fetchAnything.js')
jest.mock('../../utils/apiFetches/postUser')

describe('LoginPage', () => {
  const mockLoginFormType = 'login'
  const mockSignUpFormType = 'signup'
  const logInUrl = 'http://localhost:3000/api/users';
  const signUpUrl = 'http://localhost:3000/api/users/new';
  const mockFavoriteMovies = [{ id: 1, movie_title: 'Avengers Infinity War' }]
  const mockUserInput = {
      password: '123',
      email: 'nim@sum.com',
      confirmPassword: '123',
      userName: 'Nim'
  }
  const defaultState = {
    password: '',
    email: '',
    confirmPassword: '',
    userName: '',
    error: false,
    passwordMismatch: false
  }

  const mockCurrentUser = {
    name: '',
    id: '',
    email: '',
    favorites: []
  }

  fetchAnything.mockImplementation(() => ({ data: mockFavoriteMovies }))  

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      < LoginPage 
        currentUser={mockCurrentUser}  />
    )
  })
  it('should have set default state', () => {
    expect(wrapper.state()).toEqual(defaultState);
  })

  it('should match component snapshot with login formtype', () => {
    wrapper = shallow(
      < LoginPage 
        currentUser={ mockCurrentUser }
        formType={ mockLoginFormType }  />
    )
    expect(wrapper).toMatchSnapshot();
  })

  it('should match component snapshot with signup formtype', () => {
    wrapper = shallow(
      < LoginPage 
        currentUser={ mockCurrentUser }
        formType={ mockSignUpFormType }  />
    )
    expect(wrapper).toMatchSnapshot();
  })

  it('should handle user body and url for posting login', () => {
    wrapper = shallow(
      < LoginPage 
        currentUser={ mockCurrentUser }
        formType={ mockLoginFormType }  />
    )
    wrapper.setState({...mockUserInput });
    const expected = {
      url: 'http://localhost:3000/api/users',
      userInfo: {
        password: '123', 
        email: 'nim@sum.com', 
      },
      email: 'nim@sum.com',
      userName: 'Nim'
    }
    expect(wrapper.instance().formTypeHandler()).toEqual(expected)
  })

  it('should handle user body and url for posting signup', () => {
    wrapper = shallow(
      < LoginPage 
        currentUser={ mockCurrentUser }
        formType={ mockSignUpFormType }  />
    )
    wrapper.setState({...mockUserInput });
    const expected = {
      url: 'http://localhost:3000/api/users/new',
      userInfo: {
        password: '123', 
        email: 'nim@sum.com', 
        name: 'Nim'
      },
      email: 'nim@sum.com',
      userName: 'Nim'
    }
    expect(wrapper.instance().formTypeHandler()).toEqual(expected)
  })

  it('should fetch users favorites when logged in', async () => {
    const mockUserId = 1;
    const result = await wrapper.instance().fetchUserFavorites(mockUserId);
    expect(fetchAnything).toHaveBeenCalledWith("http://localhost:3000/api/users/1/favorites");
    expect(result).toEqual(mockFavoriteMovies);
  })

  describe('submitUserData', () => {
    
    it('should fetch post request for user login using correct params', () => {
      wrapper = shallow(
        < LoginPage 
          currentUser={ mockCurrentUser }
          formType={ mockLoginFormType }  />
      )
      wrapper.setState({...mockUserInput });
      wrapper.instance().submitUserData();
      expect(postUser).toHaveBeenCalledWith(logInUrl, {
        password: '123', 
        email: 'nim@sum.com', 
      })
    })

    it('should fetch post request for user signup using correct params', () => {
      wrapper = shallow(
        < LoginPage 
          currentUser={ mockCurrentUser }
          formType={ mockSignUpFormType }  />
      )
      wrapper.setState({...mockUserInput });
      wrapper.instance().submitUserData();
      expect(postUser).toHaveBeenCalledWith(signUpUrl, {
        password: '123', 
        email: 'nim@sum.com', 
        name: 'Nim'
      })
    })

    it('should signInUser with successfull login', () => {
      
    })
  })
})