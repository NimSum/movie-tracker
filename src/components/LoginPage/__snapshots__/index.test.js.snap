// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should match component snapshot with login formtype 1`] = `
<main
  className="login-page"
>
  <h2>
    LOGIN
  </h2>
  <form
    className="login-form"
    onSubmit={[Function]}
  >
    <fieldset
      className="login-fieldset"
    >
      <div
        className="sign-in-inputs"
      >
        <label
          className="email-input-label input-label"
          htmlFor="email-input"
        >
          E-mail
        </label>
        <input
          autoComplete="username"
          className="form-input"
          id="email-input"
          name="email"
          onChange={[Function]}
          type="email"
        />
        <label
          className="password-input-label  input-label"
          htmlFor="password-input"
        >
          Password
        </label>
        <input
          autoComplete="current-password"
          className="form-input"
          id="password-input"
          name="password"
          onChange={[Function]}
          type="password"
        />
      </div>
      <input
        type="submit"
        value="Submit"
      />
    </fieldset>
  </form>
</main>
`;

exports[`LoginPage should match component snapshot with signup formtype 1`] = `
<main
  className="login-page"
>
  <h2>
    SIGN-UP
  </h2>
  <form
    className="login-form"
    onSubmit={[Function]}
  >
    <fieldset
      className="login-fieldset"
    >
      <div
        className="signup-form-inputs"
      >
        <label
          className="sign-up input-label"
          htmlFor="userName"
        >
          Username
        </label>
        <input
          className="sign-up form-input"
          id="userName"
          name="userName"
          onChange={[Function]}
          type="text"
        />
        <label
          className="sign-up input-label"
          htmlFor="email-input"
        >
          E-mail
        </label>
        <input
          className="sign-up form-input"
          id="email-input"
          name="email"
          onChange={[Function]}
          type="email"
        />
        <label
          className="sign-up input-label"
          htmlFor="password-input"
        >
          Password
        </label>
        <input
          className="sign-up form-input"
          id="password-input"
          name="password"
          onChange={[Function]}
          type="password"
        />
        <label
          className="sign-up input-label"
          htmlFor="confirm-password-input"
        >
          Confirm Password
        </label>
        <input
          className="sign-up form-input"
          id="confirm-password-input"
          name="confirmPassword"
          onChange={[Function]}
          type="password"
        />
      </div>
      <input
        type="submit"
        value="Submit"
      />
    </fieldset>
  </form>
</main>
`;
